source/ShopPanel.cpp

```diff
@@ -616,6 +619,50 @@ bool ShopPanel::Scroll(double dx, double dy)
 
 
 
+void ShopPanel::DoFind(const string &text)
+{
+	vector<Zone>::const_iterator it = zones.begin();
+	int bestIndex = 9999;
+	vector<Zone>::const_iterator bestItem;
+	int index;
+	for( ; it != zones.end(); ++it)
+	{
+		const Ship *ship = it->GetShip();
+		const Outfit *outfit = it->GetOutfit();
+		if(ship && HasItem(ship->TrueModelName()))
+			index = Search(ship->DisplayModelName(), text);
+		else if(outfit && HasItem(outfit->TrueName()))
+			index = Search(outfit->DisplayName(), text);
```

Comment:
We should probably create a FindItem function that ShipyardPanel and OutfitterPanel implement that gets called here, just as MapSalesPanel/MapShipyardPanel/MapOutfitterPanel do.

https://github.com/endless-sky/endless-sky/blob/1664c421cf2083c49cb9d7361cb06ae671c93d82/source/MapSalesPanel.cpp#L417-L426

https://github.com/endless-sky/endless-sky/blob/1664c421cf2083c49cb9d7361cb06ae671c93d82/source/MapShipyardPanel.cpp#L173-L189

https://github.com/endless-sky/endless-sky/blob/1664c421cf2083c49cb9d7361cb06ae671c93d82/source/MapOutfitterPanel.cpp#L164-L180

Reply:
What'd be the best way to do this? I wanted to make a FindItem function separate to `ShopPanel`, but `ShipyardPanel` and `OutfitterPanel` don't have access to any of the zones that require iterating through (which is why I left it where it currently is). I'm not sure what the best approach is from a code style perspective

Reply:
> ShipyardPanel and OutfitterPanel don't have access to any of the zones

They don't? The `zones` variable is protected, not private.

Reply:
Fyi, `protected` variables are accessible from child classes.

Reply:
Yup you were both right; I read the code wrong. I've split the functionality into the child classes now 👍